File: statistics/measureOfDispersion.R
data_set <- c(5, 2, 3, 4, 1, 6, 7, 8, 10)

# i. Range
data_range <- max(data_set) - min(data_set)
print(paste("Range =", data_range))

# ii. Variance
variance <- var(data_set)
print(paste("Variance =", variance))

# iii. Standard Deviation
standard_deviation <- sd(data_set)
print(paste("Standard Deviation =", standard_deviation))

# iv. Mean Absolute Deviation
mad_value <- mad(data_set, center = mean(data_set))
print(paste("Mean Absolute Deviation =", mad_value))

# v. Quartile Deviation
quartiles <- quantile(data_set)
q1 <- quartiles[2]  # First Quartile
q3 <- quartiles[4]  # Third Quartile
quartile_deviation <- (q3 - q1) / 2
print(paste("First Quartile (Q1) =", q1))
print(paste("Third Quartile (Q3) =", q3))
print(paste("Quartile Deviation =", quartile_deviation))

# Interquartile Range (IQR)
iqr_value <- IQR(data_set)
print(paste("Interquartile Range (IQR) =", iqr_value))

# manually
data_set <- sort(data_set)
data_set_length <- length(data_set)

# i. Range
data_range <- data_set[data_set_length] - data_set[1]
print(paste("Range =", data_range))

# ii. Mean
mean_value <- sum(data_set) / data_set_length
print(paste("Mean =", mean_value))

# iii. Variance (Manual Calculation)
variance <- 0
for (i in 1:data_set_length) {
  variance <- variance + (data_set[i] - mean_value)^2
}
variance <- variance / data_set_length
print(paste("Variance =", variance))

# iv. Standard Deviation (Manual Calculation)
standard_deviation <- sqrt(variance)
print(paste("Standard Deviation =", standard_deviation))

# v. Mean Absolute Deviation (MAD)
mean_deviation <- 0
for (i in 1:data_set_length) {
  mean_deviation <- mean_deviation + abs(data_set[i] - mean_value)
}
mean_deviation <- mean_deviation / data_set_length
print(paste("Mean Absolute Deviation =", mean_deviation))

# vi. Quartile Deviation
# First Quartile (Q1)
q1_index <- ceiling(data_set_length / 4)
q1 <- data_set[q1_index]

# Third Quartile (Q3)
q3_index <- ceiling(3 * data_set_length / 4)
q3 <- data_set[q3_index]

# Quartile Deviation
quartile_deviation <- (q3 - q1) / 2
print(paste("First Quartile (Q1) =", q1))
print(paste("Third Quartile (Q3) =", q3))
print(paste("Quartile Deviation =", quartile_deviation))

File: DSAndML/statistics/measureOfDispersion.py
import numpy as np

data_set = [5, 2, 3, 4, 1, 6, 7, 8, 10]
data_set_length = len(data_set)
data_set.sort()

# i. Range
data_range = data_set[-1] - data_set[0]
print("Range = ", data_range)

# ii. Variance
# Calculate the mean
mean = sum(data_set) / data_set_length
print("Mean = ", mean)

variance = 0
for i in range(data_set_length):
    variance += (data_set[i] - mean) ** 2  # Square the difference from the mean

variance = variance / data_set_length  # Average of the squared differences (variance)
print("Variance = ", variance)

# iii. Standard Deviation
# Calculate the standard deviation (square root of variance)
standard_deviation = variance**0.5
print("Standard Deviation = ", standard_deviation)

# iv. Mean Deviation
# Calculate the mean deviation (summation of |x - mean|) / n
mean_deviation = sum(abs(data_set[i] - mean) for i in range(data_set_length))
print("Mean Deviation = ", mean_deviation / data_set_length)

# v. Quartile Deviation (Semi-Interquartile Range)

# First Quartile (Q1)
Q1 = (
    data_set[data_set_length // 4]
    if data_set_length % 2 == 0
    else (data_set[data_set_length // 4] + data_set[(data_set_length // 4) + 1]) / 2
)

# Third Quartile (Q3)
Q3 = (
    data_set[3 * data_set_length // 4]
    if data_set_length % 2 == 0
    else (data_set[3 * data_set_length // 4] + data_set[(3 * data_set_length // 4) + 1])
    / 2
)

# Quartile Deviation
quartile_deviation = (Q3 - Q1) / 2
print("First Quartile (Q1) = ", Q1)
print("Third Quartile (Q3) = ", Q3)
print("Quartile Deviation = ", quartile_deviation)

# v. Quartile Deviation (Manual Calculation)
Q1 = np.percentile(data_set, 25)  # First Quartile (25th percentile)
Q3 = np.percentile(data_set, 75)  # Third Quartile (75th percentile)
quartile_deviation = (Q3 - Q1) / 2
print("First Quartile (Q1) = ", Q1)
print("Third Quartile (Q3) = ", Q3)
print("Quartile Deviation = ", quartile_deviation)
